core/src/main/scala/org/apache/spark/storage/DiskStore.scala

```diff
@@ -67,7 +67,7 @@ private[spark] class DiskStore(
     var threwException: Boolean = true
     try {
       writeFunc(out)
-      blockSizes.put(blockId.name, out.getCount)
+      blockSizes.put(blockId, out.getCount)
```

Comment:
Though this is equal to the previous version of code, what do we gain from making this change?

Reply:
This is in line with the other part of the PR: `name` allocates a String/StringBuilder. Why use it as a hash key and pay the price of unnecessary allocation, when `BlockId` is hashable?

Reply:
@superbobry I think the last test failure is legit as you need to update the call to remove(blockId.name) on about line 116.

I was surprised it even compiles, but, for legacy reasons the JDK collections classes don't have generic types on methods like remove, so it accepts any object. That however should be the last change here.